lib/OnyxConnectionManager.ts

```diff
@@ -0,0 +1,266 @@
+        // - `connectOptions.reuseConnection` is `false`. That means the subscriber explicitly wants the connection to not be reused.
+        // - `connectOptions.initWithStoredValues` is `false`. This flag changes the subscription flow when set to `false`, so the connection can't be reused.
+        // - `withOnyxInstance` is defined inside `connectOptions`. That means the subscriber is a `withOnyx` HOC and therefore doesn't support connection reuse.
+        if (connectOptions.reuseConnection === false || connectOptions.initWithStoredValues === false || 'withOnyxInstance' in connectOptions) {
+            suffix += `,uniqueID=${Str.guid()}`;
+        }
+
+        return `onyxKey=${connectOptions.key},initWithStoredValues=${connectOptions.initWithStoredValues ?? true},waitForCollectionCallback=${
+            connectOptions.waitForCollectionCallback ?? false
+        }${suffix}`;
+    }
+
+    /**
+     * Fires all the subscribers callbacks associated with that connection ID.
+     */
+    private fireCallbacks(connectionID: string): void {
+        const connection = this.connectionsMap.get(connectionID);
+
+        connection?.callbacks.forEach((callback) => {
+            callback(connection.cachedCallbackValue, connection.cachedCallbackKey as OnyxKey);
+        });
+    }
+
+    /**
+     * Connects to an Onyx key given the options passed and listens to its changes.
+     *
+     * @param connectOptions The options object that will define the behavior of the connection.
+     * @returns The connection object to use when calling `disconnect()`.
+     */
+    connect<TKey extends OnyxKey>(connectOptions: ConnectOptions<TKey>): Connection {
+        const connectionID = this.generateConnectionID(connectOptions);
+        let connectionMetadata = this.connectionsMap.get(connectionID);
+        let subscriptionID: number | undefined;
+
+        const callbackID = String(this.lastCallbackID++);
+
+        // If there is no connection yet for that connection ID, we create a new one.
+        if (!connectionMetadata) {
+            let callback: ConnectCallback | undefined;
+
+            // If the subscriber is a `withOnyx` HOC we don't define `callback` as the HOC will use
+            // its own logic to handle the data.
+            if (!('withOnyxInstance' in connectOptions)) {
+                callback = (value, key) => {
+                    const createdConnection = this.connectionsMap.get(connectionID);
+                    if (createdConnection) {
+                        // We signal that the first connection was made and now any new subscribers
+                        // can fire their callbacks immediately with the cached value when connecting.
+                        createdConnection.isConnectionMade = true;
+                        createdConnection.cachedCallbackValue = value;
+                        createdConnection.cachedCallbackKey = key;
+
+                        this.fireCallbacks(connectionID);
```

Comment:
I may not understand this completely, so feel free to correct me. A `callback` is defined here, that in this exact line calls all callbacks for a given `connectionID`, wouldn't it fire callbacks indefinitely? (first `callback` calls all `callbacks` that call all `callbacks` again and again 😅)

Reply:
I think there is a difference between the callback of a `subscription` and a `connection`, right?

Here the callbacks of the `connection` are triggered, but this will not trigger the (outer) callback of the `subscription`.

Reply:
Yes, the callbacks that are being fired are the ones supplied when the consumer call `Onyx.connect` (or inside `useOnyx`), while this `callback` defined here is related to the connection, that can be reused and fire multiple consumer's callbacks.